File: go_app/internal/ui/progress.go
package ui

import (
	"fmt"
	"strings"
)

// ProgressVariant represents the possible variants of a progress bar
type ProgressVariant string

const (
	// ProgressVariantPrimary represents a primary progress bar
	ProgressVariantPrimary ProgressVariant = "primary"
	// ProgressVariantSuccess represents a success progress bar
	ProgressVariantSuccess ProgressVariant = "success"
)

// Progress represents a progress component
type Progress struct {
	Value   int
	Variant ProgressVariant
}

// NewProgress creates a new progress with the given parameters
func NewProgress(value int, variant ProgressVariant) *Progress {
	// Filter variant to ensure it's valid
	allowedVariants := []ProgressVariant{ProgressVariantPrimary, ProgressVariantSuccess}
	isValidVariant := false
	for _, v := range allowedVariants {
		if variant == v {
			isValidVariant = true
			break
		}
	}
	if !isValidVariant {
		variant = ProgressVariantPrimary
	}

	// Ensure value is between 0 and 100
	if value < 0 {
		value = 0
	} else if value > 100 {
		value = 100
	}

	return &Progress{
		Value:   value,
		Variant: variant,
	}
}

// Render generates the HTML for the progress
func (p *Progress) Render() string {
	var parts []string

	// Start progress bar div
	parts = append(parts, `<div class="gl-progress-bar progress">`)

	// Add progress div
	progressClass := fmt.Sprintf("gl-progress gl-progress-bar-%s", p.Variant)
	progressStyle := fmt.Sprintf("width: %d%%;", p.Value)
	parts = append(parts, fmt.Sprintf(`<div class="%s" style="%s"></div>`, progressClass, progressStyle))

	// Close progress bar div
	parts = append(parts, "</div>")

	return strings.Join(parts, "\n")
}

File: rust_app/src/components/pajamas/progress.rs
/// Renders a Pajamas compliant progress bar
#[derive(Debug, Clone)]
pub struct ProgressComponent {
    /// The value of the progress bar (0-100)
    value: u8,
    /// The variant of the progress bar
    variant: ProgressVariant,
}

/// The variant of the progress bar
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProgressVariant {
    /// Primary variant
    Primary,
    /// Success variant
    Success,
}

impl ProgressComponent {
    /// Create a new progress component
    pub fn new(value: u8, variant: ProgressVariant) -> Self {
        Self {
            value: value.min(100), // Ensure value is between 0 and 100
            variant,
        }
    }

    /// Get the variant class
    fn variant_class(&self) -> &'static str {
        match self.variant {
            ProgressVariant::Primary => "gl-progress-bar-primary",
            ProgressVariant::Success => "gl-progress-bar-success",
        }
    }

    /// Render the progress component
    pub fn render(&self) -> String {
        format!(
            r#"<div class="gl-progress-bar progress"><div class="gl-progress {}" style="width: {}%;"></div></div>"#,
            self.variant_class(),
            self.value
        )
    }
}
